cmd/syncthing/gui.go

```diff
@@ -910,17 +904,49 @@ func mimeTypeForFile(file string) string {
 	}
 }
 
-func toNeedSlice(fs []db.FileInfoTruncated) []map[string]interface{} {
-	output := make([]map[string]interface{}, len(fs))
-	for i, file := range fs {
-		output[i] = map[string]interface{}{
-			"name":         file.Name,
-			"flags":        file.Flags,
-			"modified":     file.Modified,
-			"version":      file.Version,
-			"localVersion": file.LocalVersion,
-			"size":         file.Size(),
-		}
+func toNeedSlice(fs []db.FileInfoTruncated) []jsonDBFileInfo {
+	res := make([]jsonDBFileInfo, len(fs))
+	for i, f := range fs {
+		res[i] = jsonDBFileInfo(f)
+	}
+	return res
+}
+
+// Type wrappers for nice JSON serialization
+
+type jsonFileInfo protocol.FileInfo
+
+func (f jsonFileInfo) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]interface{}{
+		"name":         f.Name,
+		"size":         protocol.FileInfo(f).Size(),
+		"flags":        fmt.Sprintf("%#o", f.Flags),
+		"modified":     time.Unix(f.Modified, 0),
+		"localVersion": f.LocalVersion,
+		"numBlocks":    len(f.Blocks),
+		"version":      jsonVersionVector(f.Version),
+	})
+}
+
+type jsonDBFileInfo db.FileInfoTruncated
+
+func (f jsonDBFileInfo) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]interface{}{
+		"name":         f.Name,
+		"size":         db.FileInfoTruncated(f).Size(),
+		"flags":        fmt.Sprintf("%#o", f.Flags),
+		"modified":     time.Unix(f.Modified, 0),
+		"localVersion": f.LocalVersion,
+		"version":      jsonVersionVector(f.Version),
+	})
+}
+
+type jsonVersionVector protocol.Vector
+
+func (v jsonVersionVector) MarshalJSON() ([]byte, error) {
+	res := make([]string, len(v))
+	for i, c := range v {
+		res[i] = fmt.Sprintf("%d:%d", c.ID, c.Value)
```

Comment:
I'm going to vote against, because

```[JMFJC] 23:26:44 DEBUG: json: error calling MarshalJSON for type main.jsonFileInfo: json: error calling MarshalJSON for type main.jsonVersionVector: json: unsupported type: map[uint64]uint64```

(maps of int to int don't really exist in javascript apparently, because that's an array... yeah... :scream:)